### ISSUE TITLE
Configurable kube-scheduler strategies (a.k.a. bin-packing)

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener#1455]
**What would you like to be added**:
We should investigate what is necessary to allow configuring the scheduling strategies that are used by the `kube-scheduler`.

Disclaimer: @mvladev pointed out that the Kubernetes community is currently planning to change the way how the scheduler is configured:

* https://docs.google.com/document/d/1P1eBaePxuko22B8LLOrPEonoj1awc1dCVBDefuRWUA4
  * Migrate predicates/priorities to Plugins. Custom predicates/priorities will need equivalent Plugin configuration APIs.
  * Embed `ExtenderConfig` into the scheduler's component config.
* Deprecate the component config parameter that allows specifying Policy source. This should be easy since the scheduler's component config is still v1alpha1.

Consequently, it's probably not a good idea to start exposing anything *right now* as there are  drastic design changes planned. Anything that we expose in our API today must be backwards compatible, hence, our TCO in that regards will increase significantly if the design changes planned by the community are put into practice.

**Why is this needed**:
End-users want to optimize the kube-scheduler for their workloads, and the standard strategy does not always suffice.

**Comment by yarivlifchuk** [comment]
Application Studio (WebIDE DevX) need this capability to avoid creating all PODS in the same node while there are several pre allocated empty nodes. Load on one node cause to PODs failures. 

Support configurable Shoot K8S strategy scoring  - strategy for Application Studio (WebIDE) should be changed from ImageLocalityPriority to leastRequestedPriority
Currently it creates problematic scheduling behavior in which all pods are allocated on one node out of five.

**Comment by vlerenc** [comment]
@rfranzke @timuthy Given that with `v1.22` `KubeSchedulerConfiguration` in version `kubescheduler.config.k8s.io/v1beta2` is supported and not only we as in Gardener, but also other big players like to have out-of-the-box support for more scheduling strategies (in particular/specifically bin-packing), can we do something with reasonable effort?

(1) On the one hand, we could possibly offer a feature similar to the audit log configuration where stakeholders place `KubeSchedulerConfiguration` in their Gardener project and refer to it from the `shoot` spec and we inject it into their `kube-scheduler`, but that's probably so much work that it will postponed for a long time.

(2) On the other hand, we could quickly help us and these big players by offering a custom spec field (like https://github.com/gardener/gardener/blob/master/example/90-shoot.yaml#L197) for bin-packing. That should be minimal effort, basically:
- Adding this spec field
- Switching the scheduler for `1.22` clusters to configuration with `NodeResourcesFit` with either `LeastAllocated` (default) or `MostAllocated` and testing that it works as expected

(3) We could even do without a custom spec field and "simply" add another profile in our `KubeSchedulerConfiguration` from `v1.22`, e.g. `binpacking-scheduler` (as in https://kubernetes.io/docs/reference/scheduling/config/#multiple-profiles). If we give them a version, e.g. `binpacking-scheduler-v1` we could not only add more profiles over time, but also introduce new versions.

Yes, the disadvantage is, that if we later add full `kube-scheduler` configurability, we might want to deprecate that custom spec field/profile again. Then again, we don't have to if we define precedence (in the case of the custom spec field), meaning that if the full configuration is given it trumps the custom spec field above. WDYT?

P.S.: The advantage of implementing this ticket is that we could stop deploying a custom scheduler for seeds (#4617) and rather leverage this feature as a regular shoot cluster feature.

**Comment by vlerenc** [comment]
@timuthy, @ialidzhikov and I discussed the subject matter and @timuthy and I suggested in the end to introduce additional `kube-scheduler` profiles, at first only `most-allocated-node-scheduler` that configures the `NodeResourcesFit` plugin to `MostAllocated`. That should help with the most requested demand for a second scheduling profile and it will make things also simpler for us by allowing us to remove the customer seed scheduler. OK with you @ialidzhikov ?

**Comment by timuthy** [comment]
> will make things also simpler for us by allowing us to remove the customer seed scheduler.

There is one exception, though. We cannot get rid of the custom scheduler if we want to have it in soil clusters.

**Comment by vlerenc** [comment]
@timuthy True, but I was thinking, we don't really speak about lots of money here for soil clusters (or not-shooted seeds behind firewalls) as we have only few of them and every operator should try to avoid them. So, would we really keep that complexity just for that use case? Not sure it's worth it - I would drop it and reduce the complexity and accept the few bucks more we pay for the very few soils.
Also, with clustered ETCD we can be more aggressive again with cluster-autoscaler scale-in again for all worker pools.

**Comment by timuthy** [comment]
I had a discussion with @rfranzke who provided some brackground und history about the scheduler configuration behind the `SeedKubeScheduler`.

Let me outline the benefits and drawbacks of the 3 options we have:

1. Shoot owner deploys a separate Kube-Scheduler into the shoot:
    ➕ Scheduler component and config is entirely controlled by shoot owner who know his application demands the best.
    ➕ Kube-Scheduler can run with the a Kubernetes version skew of 1 ([ref](https://kubernetes.io/releases/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager)).
    ➖ Owner has to operate the Scheduler component.

2. Gardener adds a second profile (same as used for `SeedKubeScheduler`) to the Kube-Scheduler config:
    ➕ Easy to consume in shoot as it comes "for free".
    ➖ Opinionated configuration for which we don't know it suits the users' use-cases (although it is very static).
    ➖ The configuration behind `SeedKubeScheduler` is not final yet (still alpha) as we experienced some problems with it in the past and chances are high that we need further adjustments. Hence, we cannot confidentally deploy the configuration we have today to shoot clusters as well.
    ➖ Once rolled out to shoots we cannot change the configuration as these would be incompatible (maybe only with deprecation period etc.)
    ➖ It opens the gate for further profile requests.

3. Gardener adds additional profile support via the Shoot API:
    ➕ Additional profiles can be configured by users, i.e. they still have the control.
    ➖ Gardener needs to add handling for Kube-Scheduler configs in different versions (transferring fields in Shoot to Kube-Scheduler config).
    ➖ Gardener needs to add validation for configuration options. Since each Kubernetes release can add and drop plugins this will be additional effort.

Please let us iterate again and check if option 2 (https://github.com/gardener/gardener/issues/1455#issuecomment-959240827) is still the one we want to choose.

**Comment by vlerenc** [comment]
@timuthy @rfranzke Yes, whatever it's worth, I am still in favour of 2), because:
- It is easy to offer (minimal effort on our end)
- It is easy to consume (minimal effort on customer end), so it's the sweet spot
- Yes, it's opinionated, but there are not many options when it comes to `NodeResourcesFit` specifically or the subject matter in general (for instance, Swarm had only `spread`, `random`, and `bin-packing` or which `random` was garbage - Kubernetes' `ratio` makes a lot more sense and maybe we add a `ratio-80` for 80% later, but so what?)
- The strategies can be either updated with every new K8s release or new ones easily added (with version numbers, see ideas above)
- More profiles can be added, which is portrayed as disadvantage, which I get, but the generic solution seems so much more effort to implement and use - and all that not knowing whether more profiles will be requested at all (in the past one profile and only one was requested multiple times: `bin-packing`, so the evidence is that this is in high demand and not much else, if anything)
- Even if we learn that 2) is the wrong path, we can correct it with the next K8s release - to implement it is minimal effort if I am not wrong

So, I have trouble understanding the rational behind the overly negative display of option 2). Key stakeholders (chief amongst them even us) will not go away with their ask to support bin-packing. The effort to deploy your own scheduler (including us) is nothing anybody wants to do. Offering full configurability is an even wider API surface, a lot more effort, and most likely not really required, AND we can easily correct the decision later.

**Comment by timuthy** [comment]
Thanks for the points @vlerenc. Although I was very much in favor of 2) in the beginning, I tend to think that the minimal effort of 2) will be a fallacy and will put a lot of stress on dev and ops on our side because:

> Yes, it's opinionated, but there are not many options when it comes to NodeResourcesFit specifically or the subject matter in general (for instance, Swarm had only spread, random, and bin-packing or which random was garbage - Kubernetes' ratio makes a lot more sense and maybe we add a ratio-80 for 80% later, but so what?)

Being opinionated about this very crucial component will become the # 1 reason for hot-fixes. We ship a 80% ratio profile, then users will make use of it and realize that 80% breaks them. They will request a 73% profile and of course they need it by yesterday because of ... (you know the stories). In such cases I'd rather accept the higher effort in the beginning for 3) but then give users the possibility to help themselves by changing their shoot definition.

> (in the past one profile and only one was requested multiple times: bin-packing, so the evidence is that this is in high demand and not much else, if anything)

IIUC, it's not clear or tested yet if that bin-packing profile (opinionated) will really serve the needs of the parties that requested this profile. I honestly cannot imagine that we can ship a single bin-packing profile that will match the majorities' requirements -> we'll end up with hot-fixing again.

**Comment by vlerenc** [comment]
> put a lot of stress on dev and ops on our side

@timuthy I think, you are a bit too pessimistic/exaggerating here. Put that into relation to the effort spent already:
- Big players asking again and again and we repeat this conversation ad nauseam was wasteful so far - personally, I want that to end
- We implementing the custom seed scheduler instead of doing the rational thing that we always did in the past: no special treatment of us, i.e. if people need a feature and we need it, we develop one feature for all parties

So, I really dislike the effort of our custom scheduler and I dislike it even more to tell  people that are in need of the very same thing (that also we need) to do it themselves when its a component that is fundamentally part of the managed service.

> being opinionated about this very crucial component will become the # 1 reason for hot-fixes

Isn't that a bit much now? As illustrated, there is not much choice between spread and binpacking, neither in other container schedulers nor in Kubernetes (see above). The 80% was just an example and nobody says we go there. We can always tell people to deploy a customer scheduler, but if so many (including us) ask for bin-packing, why wouldn't we then offer it? I don't get the resistance given all the arguments pro and con. I understand the cons, but not the conclusion to not do it or do the complex solution for which we neither see the demand/pull nor would we have the time.

> IIUC, it's not clear or tested yet if that bin-packing profile (opinionated) will really serve the needs of the parties that requested this profile.

That's why I don't understand why we tell people to do it themselves. Valuable features, that's how we always did it, were developed for ourselves and rolled out then later to everybody. The custom seed scheduler won't be rolled out - the profile can be. Also, the discussion is too theoretical. What are the issues, why isn't it working? This needs to be resolved either way. You can even add an `_alpha` suffix to the scheduling profile if this feels better and worst case we can remove it again and implement the big thing, but I don't see the demand/pull for the big thing at all. There are more critical topics were we still live we hacks like the Terraformer and here we want to be "perfect" though we lack demand/pull? Don't we have better things to do than discussing this subject matter again and again or implementing the big thing that widens our API surface considerably and binds our dev capacity as well?

**Comment by rfranzke** [comment]
I think there is no conclusion or decision yet - we just had a chat about the potential implementations and wanted to outline our results here.

For the time being we can certainly go with option 2 and there are many reasons to do so given that we are actually busy with other topics. However, for me personally, the major concern with this approach is that we don't have experience with bin-packing ourselves, and "blindly" offering it to end-users is risky.

The last time we tried out the `SeedKubeScheduler` feature gate with this configuration (https://github.com/gardener/gardener/blob/58d136a7753df1160da613018f63b59693951df6/pkg/operation/botanist/component/gardenerkubescheduler/gardener_kube_scheduler_bootstrap.go#L66-L120) we had to disable it again since "it didn"t work". I guess, at this point nobody knows/remembers anymore what the concrete reasons were. Yes, the issues have to be solved anyways since we want to use the feature ourselves. Still, somebody should better do this first before we offer it for end-users. When we add the profile today then people will use it tomorrow. I don't want to join any bridge calls if this results in problems or outages because then it's "Gardener's scheduler who messed it up".

Anyways, I'm too far away from the details and it's all too fuzzy, I agree. Just saying that - given the bad experiences - we should rather get a better understanding of this new scheduler behaviour with our own infrastructure and clusters first before we now roll it out to everybody.

**Comment by vlerenc** [comment]
@rfranzke OK, the goal is not to implement the feature and then throw it over the fence. The goal is to implement it and use it ourselves and then give it to the stakeholders. Many will need time to upgrade until v1.22, so it's months out. And of course, if we see already issues with the plugin and it's not working as expected, it's a showstopper. Most likely though, we will have run in poor `requests` and VPA is compounding the problems as it has many deficiencies. Let's find out what these problems were and where (plugin or something else like `requests`) and then we know more to go on.

**Comment by ialidzhikov** [comment]
> The last time we tried out the `SeedKubeScheduler` feature gate with this configuration (https://github.com/gardener/gardener/blob/58d136a7753df1160da613018f63b59693951df6/pkg/operation/botanist/component/gardenerkubescheduler/gardener_kube_scheduler_bootstrap.go#L66-L120) we had to disable it again since "it didn"t work". I guess, at this point nobody knows/remembers anymore what the concrete reasons were. Yes, the issues have to be solved anyways since we want to use the feature ourselves. Still, somebody should better do this first before we offer it for end-users.

In the handover with Martin V. we talked about this issue that we hit on a large landscape. The issue was that when the `SeedKubeScheduler` feature gate was enabled, control plane Pods were still scheduled on instances that already hit their volume attachment limit on the infrastructure. We suspect that it might be an issue with the aws-ebs-csi-driver itself. Still, it might be also an issue related to our custom scheduling configuration. The issue has to be reproduced and properly investigated. This issue was actually the first thing I wanted to check when I start working on the `SeedKubeScheduler` feature gate.
There was one more concern discussed in the handover session. The bin-packing feature itself aims to improve the Node utilization by scheduling Pods to the most allocated Nodes. In our setup we don't follow the K8s recommended cgroup setup  (see https://github.com/gardener/gardener/issues/2590 for more details). The raised concern back then was that with the bin-packing enabled we may hit cgroup related issues more frequently (caused by the increased Node utilization).

**Comment by timuthy** [comment]
Thanks a lot for the insights @ialidzhikov 👍 Does it make sense to document these points and track the progress in #4741?

**Comment by ialidzhikov** [comment]
> Thanks a lot for the insights @ialidzhikov 👍 Does it make sense to document these points and track the progress in #4741?

I now updated the description of #4741 with the points that we discussed back then.

**Comment by vlerenc** [comment]
Thanks @ialidzhikov, so these are the dependencies for this ticket (more specifically, bin-packing as a feature):
- https://github.com/gardener/gardener/issues/5064
- https://github.com/gardener/gardener/issues/2590

**Comment by vlerenc** [comment]
Probably related: https://github.tools.sap/kubernetes/autoscaling/issues/39 and https://github.tools.sap/kubernetes/autoscaling/issues/40

**Comment by ialidzhikov** [comment]
Hello @vlerenc @rfranzke @timuthy ,

Having in mind that https://github.com/gardener/gardener/issues/5064 is near completion, I can start looking into this issue.
I went over the discussion in this issue to recall what were the options on the table.

I have the feeling that 2 different things are discussed as part of this issue:
1. How to allow any arbitrary scheduling profiles?
2. How to allow Shoots to benefit from the so called "bin-packing" scheduling profile.

Some thoughts on the options described in https://github.com/gardener/gardener/issues/1455#issuecomment-964267871:
> 3. Gardener adds additional profile support via the Shoot API:

- Do we mean to have our own types under `.spec.kubernetes.kubeScheduler` that describe additional scheduling profiles and that are afterwards translated to a K8s scheduling profiles? Let me know if I misunderstand this option. I cannot image how this would look like:
    - the KubeSchedulerConfiguration is not v1 or stable - they can launch a new api version (for example `kubescheduler.config.k8s.io/v1beta4`) but we cannot evolve the Shoot API in a similar way. Even if the KubeSchedulerConfiguration is v1, there is no guarantee that they won't launch `v2alpha1`.
    - to my impression the profiles can be quite verbose (ref https://kubernetes.io/docs/reference/scheduling/config/#extension-points) and does not look okay to me to allow such verbose configuration inline in the Shoot type

I also asked myself what the others do? I find that GKE has [`autoscaling profiles`](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-autoscaler). They offer to 2 autoscaling profiles - `balanced` and `optimize-utilization`. To my understanding this is very close to option 2. from https://github.com/gardener/gardener/issues/1455#issuecomment-964267871 - supporting multiple scheduling profiles: the `default` scheduling profile (balanced strategy) and the new `bin-packing` profile. Additionally, I conclude that the `optimize-utilization` autoscaling profile comes with a cluster autoscaler that is tweaked as the docs state that "the cluster autoscaler scales down the cluster more aggressively: it can remove more nodes, and remove nodes faster.". When the `optimize-utilization` profile is enabled, GKE "sets the scheduler name in the Pod spec to `gke.io/optimize-utilization-scheduler`". It seems that they went 1 step further - they don't only provide the scheduling profile but they impose it. And I think that's all - they don't allow configuring any custom scheduling profile (I assume you have to deploy your own kube-scheduler to achieve this).

---

Long story short, I tend to agree with option 2.

with the following disclaimers:
- it is alpha and use it on your own risk
- it is closed to modification requests
- no other scheduling profile requests will be acccepted
- deploy and maintain own kube-scheduler if the predefined bin-packing profile does not fit you

and the following questions:
- Do we also impose the new scheduling profile (via mutating webhook) as GKE does in the whole cluster or do we rely the end user to explicitly specify `.spec.schedulerName=<new-bin-packing-profile>` in its workload?
  - If we impose the new scheduling profile via mutating webhook then we would need a field on Shoot level that indicates which scheduling profile to be used.
- In the context of SeedKubeScheduler how would we handle the non-ManagedSeeds? Currently we have a mutating webhook for Pods as part of the seed-admission-controller that sets the schedulerName field to `gardener-kube-scheduler`. We would need the webhook only for ManagedSeeds and not for the others as they won't have this scheduling profile (the GKE soil for example).

**Comment by rfranzke** [comment]
Perhaps we should rather have a meeting to talk about the open questions? There were many discussions in the past and I cannot follow everything anymore, neither do I know whether all arguments still apply today.

**Comment by vlerenc** [comment]
@ialidzhikov Thank you for the summary. One question in regards to the web hook: Why is that needed? I thought, we can more elegantly, if the end user selects that feature in the shoot spec, configure the default scheduler accordingly. No?

**Comment by ialidzhikov** [comment]
Do you see in [the docs](https://kubernetes.io/docs/reference/scheduling/config/#multiple-profiles) a way to mark a profile as default? I didn't look into all of the details but looks like it is the kube-apiserver that sets `.spec.schedulerName` to `default-scheduler` when the field is not specified:
> If a Pod doesn't specify a scheduler name, kube-apiserver will set it to `default-scheduler`. Therefore, a profile with this scheduler name should exist to get those pods scheduled.

Things to check:
 - is the default scheduler configurable in the kube-apiserver? I think it can be a valid feature request the default scheduler name to be  configurable in the kube-apiserver 🤔 👀 
 - do GKE handle the scenario with a mutating webhook?

**Comment by ialidzhikov** [comment]
> do GKE handle the scenario with a mutating webhook?

I now checked and they do it via mutating webhook. I created a GKE cluster for test purposes and switched the autoscaling profile from `default` to `optimize-utilization`. This created the following webhook in the cluster.

```
$ k get mutatingwebhookconfigurations.admissionregistration.k8s.io mutate-scheduler-profile.config.common-webhooks.networking.gke.io
NAME                                                                WEBHOOKS   AGE
mutate-scheduler-profile.config.common-webhooks.networking.gke.io   1          12s
```
<details>

```yaml
apiVersion: admissionregistration.k8s.io/v1
kind: MutatingWebhookConfiguration
metadata:
  labels:
    networking.gke.io/common-webhooks: "true"
webhooks:
- admissionReviewVersions:
  - v1beta1
  clientConfig:
    caBundle: "<omitted>"
    url: https://localhost:5443/webhook/mutate-scheduler-profile
  failurePolicy: Ignore
  matchPolicy: Equivalent
  name: mutate-scheduler-profile.common-webhooks.networking.gke.io
  namespaceSelector: {}
  objectSelector: {}
  reinvocationPolicy: Never
  rules:
  - apiGroups:
    - ""
    apiVersions:
    - v1
    operations:
    - CREATE
    resources:
    - pods
    scope: '*'
  sideEffects: None
  timeoutSeconds: 10
```

</details>

---

For the time being we can start with similar mutating webhook and in parallel work on making the default scheduler name configurable in the kube-apiserver (ref https://github.com/kubernetes/kubernetes/blob/f4abde9e57cc30efcc8f79c9ebd791d15f4164c4/pkg/apis/core/v1/defaults.go#L205-L207).

**Comment by ialidzhikov** [comment]
Outcomes from the meeting we had earlier today:

- I think we are okay to try out option 2. from https://github.com/gardener/gardener/issues/1455#issuecomment-964267871 with the disclaimers from https://github.com/gardener/gardener/issues/1455#issuecomment-1163330012:
- We plan to introduce a new field to the Shoot spec that denotes the kube-scheduler profile that is used/imposed
  ```yaml
  spec:
    kubernetes:
      kubeScheduler:
        profile: "default" # (allowed values "default" and "bin-packing")
  ```

- We need to impose the profile. Otherwise the user has to explicitly set the `.spec.schedulerName` in its manifests which is not a good experience. The options on the table are:
  - Maintain 2 profiles in the scheduler's component config and use the mutating webhook approach similar to GKE (ref https://github.com/gardener/gardener/issues/1455#issuecomment-1166597905).
    - (+) While debugging issues it makes it clear to the end user/operator that a non-default scheduling profile is used.
    - (+) The `default-scheduler` is still preserved and can be used by the end user.
    - (-) It comes with overhead that we need to maintain new repo/image for the webhook and deploy it as part of the control plane.
  - Always maitain 1 profile named `default-scheduler` in the scheduler's component config and configure it according to the profile requested in the Shoot spec.
    - (-) Not transparent to the end user/operator that actually the bin-packing scheduling profile is used.
    - (-) The `default-scheduler` is not preserved and cannot be used.
    - (+) Straightforward to implement.
  - (only longterm) Push for https://github.com/kubernetes/kubernetes/issues/110795.

- We need check what needs to be done to consider also GPU weight in the bin packing profile.

- Due to https://github.com/gardener/gardener/issues/5064 the bin packing profile will be allowed only for K8s >= 1.20.

- The `SeedKubeScheduler` feature will be deprecated and dropped in favour of the new bin packing profile.


@vlerenc @timuthy @rfranzke thank you for the help and the joint discussion!

**Comment by vlerenc** [comment]
Sounds good, thank you @ialidzhikov. Just two things:
- I would suggest to call the default profile (in the shoot spec) rather `balanced`, but not `default` as that carries not enough meaning and most people don't know what the default is
- I would add always all scheduler profiles with explicit names, i.e. for now only `balancedScheduler` and `binpackingScheduler` (maybe later more), whether we use the web hook (1) or the overwrite (2) approach (the latter being my favorite as I think the disadvantages are not as serious since, see above, most people don't know what the default is and the advantage is that we end up with fewer components/web hooks that must be maintained and that can fail)

**Comment by ialidzhikov** [comment]
> * I would suggest to call the default profile (in the shoot spec) rather `balanced`, but not `default` as that carries not enough meaning and most people don't know what the default is

Ok, let's use `balanced`.

> * I would add always all scheduler profiles with explicit names, i.e. for now only `balancedScheduler` and `binpackingScheduler` (maybe later more)

I was about to use `default-scheduler` (to keep the defaulting from kube-apiserver side working) and `bin-packing-scheduler` - so yes, we will use explicit and descriptive names.

> whether we use the web hook (1) or the overwrite (2) approach

For now I will pick the webhook approach due to the advantages listed above; this is also what GKE does; and I, myself, as operator/end user would like to see a schedulerName that corresponds to what it actually does.
